Support for Dnsmasq

beinvisible opened the pull request:
Dnsmasq provides DHCP and TFTP functionality; it also acts as DNS and NTP proxy.


To enable set 'dhcp server' in provision.conf to 'dnsmasq'.

The config is added to /etc/dnsmasq.d/ if this directory exists, else the global /etc/dnsmasq.conf is overwritten.
The location of this file can also be configured via the 'dhcp config file' in provision.conf.

This patch also adds symlinks to the PXE boot images which include ".0" in the filename. This is because some PXE implementations seem to be hardcoded to use "image.0" when booting (e.g. on the Cisco machines I work with).

bensallen commented on provision/etc/dnsmasq-template.conf:
```diff
@@ -0,0 +1,21 @@
+pxe-service=tag:!ipxe,X86-64_EFI,"EFI x86-64",/warewulf/ipxe/bin-x86_64-efi/snp.efi
+pxe-service=tag:!ipxe,BC_EFI,"EFI BC",/warewulf/ipxe/bin-x86_64-efi/snp.efi
+pxe-service=tag:!ipxe,IA32_EFI,"EFI IA32",/warewulf/ipxe/bin-i386-efi/snp.efi
+pxe-service=tag:!ipxe,x86PC,"x86 PC",/warewulf/ipxe/bin-i386-pcbios/undionly.kpxe
```
Is there an aarch64 rule we can use here?

beinvisible replied:
According to dnsmasq's manpage: "The known types are x86PC, PC98, IA64_EFI, Alpha, Arc_x86, Intel_Lean_Client, IA32_EFI, X86-64_EFI, Xscale_EFI, BC_EFI, ARM32_EFI and ARM64_EFI; an integer may be used for other types."

Guess that would be ARM64_EFI then...

bensallen replied:
Agreed ARM64_EFI seems correct.

bensallen replied:
Curious the difference between dhcp-boot and pxe-service options. Seemingly from the man page we should be use dhcp-boot instead of pxe-server.

beinvisible replied:
Yes. It's a bit more complicated because the system type has to extracted manually, but it solves the problem with the .0 suffix.

bensallen replied:
Awesome.

bensallen commented on provision/lib/Warewulf/Provision/Pxe.pm:
```diff
@@ -100,6 +100,9 @@ setup()
                 }
             }
+            if (-f "$tftpdir/warewulf/ipxe/$f") {
+                system("ln -sf $tftpdir/warewulf/ipxe/$f $tftpdir/warewulf/ipxe/$f.0");
```
I'd prefer we not bake this `*.0` specific fix into Warewulf. This workaround seems like something that should be put into site location configuration management.

beinvisible replied:
Maybe just a "boot image suffix" setting in provision.conf which is used when copying the images?

bensallen replied:
Even that seems like something that shouldn't be in Warewulf.

beinvisible replied:
Well, the boot images are managed/provided by WW. The problem with my machines is, that they explicitly request "image.0" when I configure "image". Isc-dhcp seems to just try to find the file without the suffix, but I haven't figured out yet how to do that with dnsmasq.

bensallen replied:
From the dnsmasq main page for pxe-service: ```Note that the "layer" suffix (normally ".0") is supplied by PXE, and need not be added to the basename. Alternatively, the basename may be a filename, complete with suffix, in which case no layer suffix is added.```

Can you try using dhcp-boot instead of pxe-service?

bensallen commented on cluster/libexec/wwinit/50-tftp.init:
```diff
@@ -25,4 +25,11 @@ if wwpackage_check tftp-server tftpd tftpd-hpa; then
     fi
 fi
+
+if wwpackage_check dnsmasq; then
```
We should only check for the dnsmasq package if its selected in provision.conf. Also vice versa, ISC should only be checked for it its selected. See https://github.com/warewulf/warewulf3/blob/development/common/libexec/wwinit/10-database.init#L143 for an example of how we deal with this for database types.

bensallen commented on provision/etc/Makefile.am:
```diff
@@ -13,7 +13,7 @@ endif
 httpdconfdir = $(httpdir)
 
-dist_conf_DATA = provision.conf livesync.conf dhcpd-template.conf
+dist_conf_DATA = provision.conf livesync.conf dhcpd-template.conf dnsmasq-template.conf
```
dnsmasq_template.conf isn't part of the RPM spec file: 

```
Checking for unpackaged file(s): /usr/lib/rpm/check-files /root/rpmbuild/BUILDROOT/warewulf-provision-3.8.1-0.c6ff5cd.el7.x86_64
error: Installed (but unpackaged) file(s) found:
   /etc/warewulf/dnsmasq-template.conf


RPM build errors:
    Installed (but unpackaged) file(s) found:
   /etc/warewulf/dnsmasq-template.conf
Exited with code 1
```

bensallen commented on provision/etc/Makefile.am:
```diff
@@ -13,7 +13,7 @@ endif
 httpdconfdir = $(httpdir)
 
-dist_conf_DATA = provision.conf livesync.conf dhcpd-template.conf
+dist_conf_DATA = provision.conf livesync.conf dhcpd-template.conf dnsmasq-template.conf
```
Additionally, I think we should come up with a cleaner way than installing both dhcpd-template.conf and dnsmasq-template.conf into /etc/warewulf. Having both to me seems like it will lead to a confusing user experience. Ideas?

beinvisible replied:
In an ideal world users should never have to edit those templates, so /usr/share/warewulf/ may be a better location.

bensallen replied:
Hmm. Suppose for now let's just stick them both in /etc/warewulf so we're not changing behavior for ISC.

bensallen commented on provision/lib/Warewulf/Provision/Dhcp/Dnsmasq.pm:
```diff
@@ -0,0 +1,379 @@
+        if (! $self->get("FILE")) {
+            &dprint("Probing dhcpd looking for a default config path\n");
+            if (-x "/usr/sbin/dnsmasq") {
+                open(CONF, "strings /usr/sbin/dnsmasq | grep '/dnsmasq.conf' | grep '^/etc/' |");
```
Let's not go as far running strings on the binary to auto-discover the path for dnsmasq.conf.

bensallen replied:
Instead lets just bail out at this point.

beinvisible replied:
Well, that was a copy from the Isc module. I feel that to be quite "brutal" there as well...

bensallen replied:
Ah wonderful... Well let's not repeat what we did for ISC.

bensallen reviewed (changes requested):
Please move this PR to be against the development branch instead of master.

beinvisible commented:
> Please move this PR to be against the development branch instead of master.

Ok, will do that tomorrow (need some rebasing, else it would pull in some other commits).

beinvisible committed:
Dnsmasq provides DHCP and TFTP functionality; it also acts as DNS and NTP proxy.

To enable set 'dhcp server' in provision.conf to 'dnsmasq'.

The config is added to /etc/dnsmasq.d/ if this directory exists, else the global /etc/dnsmasq.conf is overwritten.
The location of this file can also be configured via the 'dhcp config file' in provision.conf.

bensallen commented:
I think this all looks good now. Anything else before we merge?

beinvisible commented:
On 06.03.2019 17:44, Ben Allen wrote:
>
> I think this all looks good now. Anything else before we merge?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/warewulf/warewulf3/pull/178#issuecomment-470182355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/As-LPmFtkg4ffTgTcQCjtPj1uXR1smNMks5vT_BmgaJpZM4bfMFj>.
>
Not in this particular PR :-)

-- 
cu                     www: http://invisible.priv.at
  Stephan             mail: stephan@invisible.priv.at

bensallen commented:
@beinvisible Ok awesome, thanks for this PR. Looking forward to any future PRs.
